runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/middleware/FlexibleChecksumsRequestMiddleware.kt

```diff
@@ -0,0 +1,117 @@
+import kotlinx.coroutines.*
+import kotlin.coroutines.coroutineContext
+
+/**
+ * Mutate a request to enable flexible checksums.
+ *
+ * If the checksum will be sent as a header, calculate the checksum.
+ *
+ * Otherwise, if it will be sent as a trailing header, calculate the checksum as asynchronously as the body is streamed.
+ * In this case, a [LazyAsyncValue] will be added to the execution context which allows the trailing checksum to be sent
+ * after the entire body has been streamed.
+ *
+ * @param checksumAlgorithmName the name of the algorithm used to calculate the checksum
+ */
+@InternalApi
+public class FlexibleChecksumsRequestMiddleware(private val checksumAlgorithmName: String) : ModifyRequestMiddleware {
+    public override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
+        val logger = coroutineContext.getLogger<FlexibleChecksumsRequestMiddleware>()
+
+        val headerName = "x-amz-checksum-${checksumAlgorithmName.lowercase()}"
+        logger.debug { "Resolved header name: $headerName" }
+
+        // remove all checksum headers except for $headerName
+        // this handles the case where a user inputs a precalculated checksum, but it doesn't match the input checksum algorithm
+        req.subject.headers.removeAllChecksumHeadersExcept(headerName)
+
+        val checksumAlgorithm = checksumAlgorithmName.toHashFunction() ?: throw ClientException("Could not parse checksum algorithm $checksumAlgorithmName")
+
+        if (!checksumAlgorithm.isSupported) {
+            throw ClientException("Checksum algorithm $checksumAlgorithmName is not supported for flexible checksums")
+        }
+
+        if (req.subject.body.isEligibleForAwsChunkedStreaming) {
+            req.subject.header("x-amz-trailer", headerName)
+
+            val deferredChecksum = CompletableDeferred<String>()
+
+            if (req.subject.headers[headerName] != null) {
+                logger.debug { "User supplied a checksum, skipping asynchronous calculation" }
+
+                val checksum = req.subject.headers[headerName]!!
+                req.subject.headers.remove(headerName) // remove the checksum header because it will be sent as a trailing header
+
+                deferredChecksum.complete(checksum)
+            } else {
+                logger.debug { "Calculating checksum asynchronously" }
+
+                req.subject.body = req.subject.body
+                    .toHashingBody(checksumAlgorithm, req.subject.body.contentLength)
+                    .toCompletingBody(deferredChecksum, req.subject.body.contentLength)
+            }
+
+            req.subject.trailingHeaders.append(headerName, deferredChecksum)
+        } else if (req.subject.headers[headerName] == null) {
+            logger.debug { "Calculating checksum" }
+
+            val checksum = req.subject.body.readAll()?.hash(checksumAlgorithm)?.encodeBase64String()
+                ?: throw RuntimeException("Failed to calculate checksum")
```

Comment:
This `readAll()` will only be called if the body is ineligible for upload with `aws-chunked` content encoding, which means it is smaller than 1MB.

Reply:
It means the content size _is known_ and is smaller than 1MB. If `contentLength` is `null` this branch will trigger right?

Reply:
Yes you're right. What are some cases when `contentLength` could be `null`? I don't think I've ever come across it before during testing

Reply:
I don't know of many concrete use cases but [our interface defines `contentLength` as nullable](https://github.com/awslabs/smithy-kotlin/blob/main/runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/HttpBody.kt#L16-L20). In theory, I could imagine accepting as an HTTP body a stream coming from some source where the total length is unknown because it is being calculated on the fly. If it's very large, this seems like it could cause an OOM.

Are we sure the logic for `isEligibleForAwsChunkedStreaming` is correct? Should we not stream a body where the `contentLength` is unknown?

Reply:
Discussed offline, I added a rolling hash to be used when the `contentLength` is `null` but the body is replayable. 

The unsupported case is when the `contentLength` is `null` but the body is **non**-replayable -- in this case, we can't compute the checksum using a rolling hash because the body needs to be replaced after consumption. So, the SDK will just read the entire body of unknown-length and hope it does not cause an OOM error.